README.md
# Welcome to your addon

1. You might want to move this folder into the kodi addon folder for convinience when debugging. It might also be needed to be `enabled` inside of the kodi addon browser.
2. Now start coding! Just open up the `.py` file in this folder and create what you would like Kodi to do! If you're creating a plugin, please check out [this kodi routing framework](https://github.com/tamland/kodi-plugin-routing) and copy a version of that module to your kodi addon folder.
3. Write some tests, maybe? We've created a test folder for that, otherwise just delete it ;)
4. You might want to look at your `addon.xml` it should already be filled, but you will need to understand what your doing and might want to fill in some more info. So read up [here](http://kodi.wiki/view/Addon.xml).
5. Do you want some settings for your addon? Check the `settings.xml` in the resources folder. And read up [here](http://kodi.wiki/view/Settings.xml).
6. Read [this info](http://kodi.wiki/view/Add-on_structure#icon.png) and drop an icon for your addon into the `resource` folder and name it `icon.png`.
7. Read [this](http://kodi.wiki/view/Add-on_structure#fanart.jpg) and drop a addon background into the `resource` folder and name it `fanart.jpg`.
8. End up with a beautiful Kodi addon! Good for you :) Maybe you want to [share it with us](http://kodi.wiki/view/Submitting_Add-on_updates_on_Github)?

### Debugging
To get the debug logging to work, just set the global kodi logging to true and the debug logging in your addons settings.

changelog.txt
v0.2.0
- Manual search

v0.1.8
- Search for subtitles in selected languages in settings

v0.1.2
- First working version. Download subtitles from addic7ed.com

v0.0.1
- Initial version


tests/test_addicted.py
# -*- coding: utf-8 -*-
"""Docstring"""

import sys, os
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../'))
from resources.lib.addictedutils import get_show_id

class ITTests(unittest.TestCase):
    """Docstring"""

    def test_get_show_id(self):
        """Docstring"""
        suits_show_id = get_show_id("Suits")
        ds9_show_id = get_show_id("Star Trek: Deep Space Nine")
        flash_show_id = get_show_id("The Flash (2014)")
        monty_show_id = get_show_id("Monty Python's Flying Circus")
        wtf_show_id = get_show_id("Blå ögon")

        self.assertEqual("1667", suits_show_id)
        self.assertEqual("692", ds9_show_id)
        self.assertEqual("4616", flash_show_id)
        self.assertEqual("6024", monty_show_id)
        self.assertEqual("5158", wtf_show_id)


addon.py
import os
import shutil
import sys
import urllib2
import xbmc
import xbmcaddon
import xbmcgui,xbmcplugin
import xbmcvfs
import uuid
import json
import socket

import requests
from bs4 import BeautifulSoup
from bs4 import SoupStrainer
from bs4 import UnicodeDammit

__addon__       = xbmcaddon.Addon()
__scriptid__    = __addon__.getAddonInfo('id')

__profile__     = xbmc.translatePath(__addon__.getAddonInfo('profile')).decode('utf-8')
__tempfolder__  = xbmc.translatePath(os.path.join(__profile__, 'temp', '')).decode('utf-8')

__baseURL__     = "http://www.addic7ed.com"

def get_language_code(language):
  switcher = {
    "English":    "1",
    "Spanish":    "4",
    "Italian":    "7",
    "French":     "8",
    "Portoguese": "9",
    "Turkish":    "16",
    "Dutch":      "17",
    "Russian":    "19",
    "Romanian":   "26",
    "Greek":      "27",
    "Bulgarian":  "35",
    "Arabic":     "38"
  }
  return switcher.get(language, '')

def build_language_code_string(languages):
  code_string = "|"
  for language in languages:
    code = get_language_code(language)
    if code != "":
      code_string += code + "|"
  if code_string == "|":
    code_string = "|1|"
  return code_string

def get_params(string_params):
	# Remove first ? from params
	string_params = string_params[1:]
	# Create dictionary with the params
	params = {}
	params = (q.split('=') for q in string_params.split('&'))

	params_in_dictionary = dict((key, value) for (key, value) in params)
	return params_in_dictionary

def get_details_from_player():
	# Create request
	request = {
		"jsonrpc": "2.0",
		"method": "Player.GetItem",
		"params": {
			"properties": ["showtitle", "season", "episode", "duration", "file"],
			"playerid": 1
		},
		"id": "VideoGetItem"
	}
	# Make request
	response = xbmc.executeJSONRPC(json.dumps(request))
	print(response)
	parsed_response = json.loads(response)['result']['item']
	return parsed_response

def get_show_id(showName):
	# Get source
	source = urllib2.urlopen(__baseURL__).read()

	# Trim source to avoid bad parsing
	stringStart = "<span id=\"qssShow\">"
	stringEnd = "<span id=\"qsSeason\">"
	tagIndexStart = source.find(stringStart)
	tagIndexEnd = source.find(stringEnd)
	source = source[tagIndexStart:tagIndexEnd]

	# Make the soup
	soup = BeautifulSoup(source, "html.parser")
	showTag = soup.find('option', text=showName)

	return showTag['value']

def build_show_url(showId, showSeason, language_code_string):
	# show_url = __baseURL__ + "/show/" + showId + "/season/" + showSeason
  show_url = __baseURL__ + "/ajax_loadShow.php?" + "show=" + str(showId) + "&season=" + str(showSeason) + "&langs=" + language_code_string
  # print(show_url)
  return show_url

def subs_array(showURL, showDetails):
  # Get source
  source = urllib2.urlopen(showURL).read()

  # Make the soup
  soup = BeautifulSoup(source, "html.parser")
  # print(soup)
  # Iterate through rows of first table
  # and generate array with subs details
  subs = []
  for row in soup.find('table').tbody.findAll('tr'):
    if (not row.has_attr("height")) and (row.contents[1].text == str(showDetails['episode'])):
      if row.contents[7].text.encode('UTF-8') != u'':
        hi = 'true'
      else:
        hi = 'false'

      sub = {
      "season": 		    row.contents[0].text,
      "episode": 		    row.contents[1].text,
      "episodeTitle":	  row.contents[2].contents[0].text,
      "showTitle":	    showDetails["showtitle"],
      "lang":			      row.contents[3].text,
      "version":		    row.contents[4].text,
      "hi":			        hi,
      "link":			      row.find_all('a')[1].get('href'),
      }
      subs.append(sub.copy())
  return subs

def decode_html(html_string):
  converted = UnicodeDammit(html_string)
  if not converted.unicode_markup:
    raise UnicodeDecodeError(
    "Failed to detect encoding, tried [%s]",
    ', '.join(converted.tried_encodings))
  # print converted.original_encoding
  return converted.unicode_markup


def episodes_array(searchTerm):
  searchURL = __baseURL__ + "/search.php?" + "search=" + str(searchTerm) + "&Submit=Search"
  print(searchURL)
  # Get source with the search results
  source = requests.get(searchURL).text.encode('utf8')
  # Trim source to avoid bad parsing
  stringStart = "<table class=\"tabel\" align=\"center\" width=\"80%\" border=\"0\">"
  stringEnd = "<!-- table footer -->"
  tagIndexStart = source.decode('utf8').find(stringStart)
  tagIndexEnd = source.decode('utf8').find(stringEnd)
  trimmedSource = source[tagIndexStart:tagIndexEnd]
  # Make the soup
  soup = BeautifulSoup(trimmedSource, 'html.parser')
  # Get episodes from a-tags in table
  episodes = []
  episodes_table = soup.find('table')
  episode_links = episodes_table.find_all('a')
  for link in episode_links:
    episode = {
      'name': link.text,
      'href': link['href']
    }
    episodes.append(episode.copy())
  return episodes

def create_subs_list(subs):
  for sub in subs:
    filename = "[" + str(sub["version"].encode('utf8')) + "]" + str(sub["showTitle"].encode('utf8')) + " - " + str(sub["episodeTitle"].encode('utf8')) + " " + "S" + str(sub["season"]) + "E" + str(format(int(sub["episode"]), "02d"))
    listitem = xbmcgui.ListItem(label = sub["lang"],
    							label2	= filename,
    							iconImage = "",
    							thumbnailImage = xbmc.convertLanguage(sub['lang'], xbmc.ISO_639_1))
    listitem.setProperty("hearing_imp", sub["hi"])
    url = u'plugin://' + __scriptid__ + u'/?action=download&link=' + sub["link"] + u'&filename=' + filename.decode('utf8')
    # url = u'plugin://%s/?action=download&link=%s&filename=%s' % (__scriptid__.encode('utf8'), sub["link"].encode('utf8'), str(filename.encode('utf8')))
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]),url=url,listitem=listitem,isFolder=False)

def create_episodes_list(episodes):
  listitem = xbmcgui.ListItem(label2 = "Select an episode to view its subtitles.")
  xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url="", listitem=listitem, isFolder=True)
  for episode in episodes:
    listitem = xbmcgui.ListItem(label2 = episode['name'])
    # url = u'plugin://' + __scriptid__ + u'/?action=manualsearch2&link=' + sub["link"] + u'&episodename=' + episode['name'].decode('utf8') + u'&epidodelink=' + episode['link'].decode['utf8']
    url = ""
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=url, listitem=listitem, isFolder=True)

def get_sub(link):
  url = __baseURL__ + link
  rheaders = {
  'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
  'Accept-Encoding': 'gzip, deflate, sdch',
  'Connection': 'keep-alive',
  'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2715.0 Safari/537.36',
  'Host': 'www.addic7ed.com',
  'Referer': 'http://www.addic7ed.com'}
  r = requests.get(url, headers = rheaders)
  return r.text

def download(link, filename):
  sub_file = []

  if xbmcvfs.exists(__tempfolder__):
      shutil.rmtree(__tempfolder__)
  xbmcvfs.mkdirs(__tempfolder__)

  file = os.path.join(__tempfolder__, filename.decode('utf8'))
  dfile = get_sub(link)

  file_handler = open(file, "wb")
  file_handler.write(dfile.encode('UTF-8'))
  file_handler.close

  sub_file.append(file)

  return sub_file

params = get_params(sys.argv[2])
for x in params:
	print(x + ':' + params[x].replace("%2c", ","))

if params["action"] == "search":
  languages = params["languages"].replace("%2c", ",").split(",")
  language_code_string = build_language_code_string(languages)
  show_details = get_details_from_player()
  show_id = get_show_id(show_details['showtitle'])
  showURL = build_show_url(show_id, show_details['season'], language_code_string)
  subs = subs_array(showURL, show_details)
  create_subs_list(subs)
elif params["action"] == "download":
  fname = params["filename"].replace(" ", "_") + ".srt"
  subs = download(params["link"], fname)
  for sub in subs:
    listitem = xbmcgui.ListItem(label=sub)
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=sub, listitem=listitem, isFolder=False)
elif params["action"] == "manualsearch":
  episodes = episodes_array(params['searchstring'])
  create_episodes_list(episodes)


xbmcplugin.endOfDirectory(int(sys.argv[1]))



addon_old.py
import os
import shutil
import sys
import urllib2
import xbmc
import xbmcaddon
import xbmcgui,xbmcplugin
import xbmcvfs
import uuid
import json
import socket

import requests
from bs4 import BeautifulSoup
from bs4 import SoupStrainer
from bs4 import UnicodeDammit

__addon__       = xbmcaddon.Addon()
__scriptid__    = __addon__.getAddonInfo('id')

__profile__     = xbmc.translatePath(__addon__.getAddonInfo('profile')).decode('utf-8')
__tempfolder__  = xbmc.translatePath(os.path.join(__profile__, 'temp', '')).decode('utf-8')

__baseURL__     = "http://www.addic7ed.com"

def get_language_code(language):
  switcher = {
    "English":    "1",
    "Spanish":    "4",
    "Italian":    "7",
    "French":     "8",
    "Portoguese": "9",
    "Turkish":    "16",
    "Dutch":      "17",
    "Russian":    "19",
    "Romanian":   "26",
    "Greek":      "27",
    "Bulgarian":  "35",
    "Arabic":     "38"
  }
  return switcher.get(language, '')

def build_language_code_string(languages):
  code_string = "|"
  for language in languages:
    code = get_language_code(language)
    if code != "":
      code_string += code + "|"
  if code_string == "|":
    code_string = "|1|"
  return code_string

def get_params(string_params):
	# Remove first ? from params
	string_params = string_params[1:]
	# Create dictionary with the params
	params = {}
	params = (q.split('=') for q in string_params.split('&'))

	params_in_dictionary = dict((key, value) for (key, value) in params)
	return params_in_dictionary

def get_details_from_player():
  # Create request
  request = {
    "jsonrpc": "2.0",
    "method": "Player.GetItem",
    "params": {
      "properties": ["showtitle", "season", "episode", "duration", "file"],
      "playerid": 1
    },
    "id": "VideoGetItem"
  }
  # Make request
  response = xbmc.executeJSONRPC(json.dumps(request))
  print(response)
  parsed_response = json.loads(response)['result']['item']
  print("---------")
  print(parsed_response)
  print("---------")
  return parsed_response

def get_show_id(showName):
	# Get source
	source = urllib2.urlopen(__baseURL__).read()

	# Trim source to avoid bad parsing
	stringStart = "<span id=\"qssShow\">"
	stringEnd = "<span id=\"qsSeason\">"
	tagIndexStart = source.find(stringStart)
	tagIndexEnd = source.find(stringEnd)
	source = source[tagIndexStart:tagIndexEnd]

	# Make the soup
	soup = BeautifulSoup(source, "html.parser")
	showTag = soup.find('option', text=showName)

	return showTag['value']

def build_show_url(showId, showSeason, language_code_string):
	# show_url = __baseURL__ + "/show/" + showId + "/season/" + showSeason
  show_url = __baseURL__ + "/ajax_loadShow.php?" + "show=" + str(showId) + "&season=" + str(showSeason) + "&langs=" + language_code_string
  # print(show_url)
  return show_url

def subs_array(showURL, showDetails):
  # Get source
  source = urllib2.urlopen(showURL).read()

  # Make the soup
  soup = BeautifulSoup(source, "html.parser")
  # print(soup)
  # Iterate through rows of first table
  # and generate array with subs details
  subs = []
  for row in soup.find('table').tbody.findAll('tr'):
    if (not row.has_attr("height")) and (row.contents[1].text == str(showDetails['episode'])):
      if row.contents[7].text.encode('UTF-8') != u'':
        hi = 'true'
      else:
        hi = 'false'

      sub = {
      "season": 		    row.contents[0].text,
      "episode": 		    row.contents[1].text,
      "episodeTitle":	  row.contents[2].contents[0].text,
      "showTitle":	    showDetails["showtitle"],
      "lang":			      row.contents[3].text,
      "version":		    row.contents[4].text,
      "hi":			        hi,
      "link":			      row.find_all('a')[1].get('href'),
      }
      subs.append(sub.copy())
  return subs

def decode_html(html_string):
  converted = UnicodeDammit(html_string)
  if not converted.unicode_markup:
    raise UnicodeDecodeError(
    "Failed to detect encoding, tried [%s]",
    ', '.join(converted.tried_encodings))
  # print converted.original_encoding
  return converted.unicode_markup


def episodes_array(searchTerm):
  searchURL = __baseURL__ + "/search.php?" + "search=" + str(searchTerm) + "&Submit=Search"
  print(searchURL)
  # Get source with the search results
  source = requests.get(searchURL).text.encode('utf8')
  # Trim source to avoid bad parsing
  stringStart = "<table class=\"tabel\" align=\"center\" width=\"80%\" border=\"0\">"
  stringEnd = "<!-- table footer -->"
  tagIndexStart = source.decode('utf8').find(stringStart)
  tagIndexEnd = source.decode('utf8').find(stringEnd)
  trimmedSource = source[tagIndexStart:tagIndexEnd]
  # Make the soup
  soup = BeautifulSoup(trimmedSource, 'html.parser')
  # Get episodes from a-tags in table
  episodes = []
  episodes_table = soup.find('table')
  episode_links = episodes_table.find_all('a')
  for link in episode_links:
    episode = {
      'name': link.text,
      'href': link['href']
    }
    episodes.append(episode.copy())
  return episodes

def create_subs_list(subs):
  for sub in subs:
    filename = "[" + str(sub["version"].encode('utf8')) + "]" + str(sub["showTitle"].encode('utf8')) + " - " + str(sub["episodeTitle"].encode('utf8')) + " " + "S" + str(sub["season"]) + "E" + str(format(int(sub["episode"]), "02d"))
    listitem = xbmcgui.ListItem(label = sub["lang"],
    							label2	= filename,
    							iconImage = "",
    							thumbnailImage = xbmc.convertLanguage(sub['lang'], xbmc.ISO_639_1))
    listitem.setProperty("hearing_imp", sub["hi"])
    url = u'plugin://' + __scriptid__ + u'/?action=download&link=' + sub["link"] + u'&filename=' + filename.decode('utf8')
    # url = u'plugin://%s/?action=download&link=%s&filename=%s' % (__scriptid__.encode('utf8'), sub["link"].encode('utf8'), str(filename.encode('utf8')))
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]),url=url,listitem=listitem,isFolder=False)

def create_episodes_list(episodes, languages, preferredlanguage):
  listitem = xbmcgui.ListItem(label2 = "Select an episode to view its subtitles.")
  xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url="", listitem=listitem, isFolder=True)
  for episode in episodes:
    listitem = xbmcgui.ListItem(label2 = episode['name'])
    url = u'plugin://' + __scriptid__ + u'/?action=search' + u'&episodename=' + episode['name'] + u'&episodelink=' + episode['href'] + u'&languages=' + languages + u'&preferredlanguage=' + preferredlanguage + u'&aftermanual=true'
    # url = ""
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=url, listitem=listitem, isFolder=False)

def get_sub(link):
  url = __baseURL__ + link
  rheaders = {
  'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
  'Accept-Encoding': 'gzip, deflate, sdch',
  'Connection': 'keep-alive',
  'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2715.0 Safari/537.36',
  'Host': 'www.addic7ed.com',
  'Referer': 'http://www.addic7ed.com'}
  r = requests.get(url, headers = rheaders)
  return r.text

def get_details_from_episode_url(url):
  url_in_array = url.split("/")
  return url_in_array

def download(link, filename):
  sub_file = []

  if xbmcvfs.exists(__tempfolder__):
      shutil.rmtree(__tempfolder__)
  xbmcvfs.mkdirs(__tempfolder__)

  file = os.path.join(__tempfolder__, filename.decode('utf8'))
  dfile = get_sub(link)

  file_handler = open(file, "wb")
  file_handler.write(dfile.encode('UTF-8'))
  file_handler.close

  sub_file.append(file)

  return sub_file

params = get_params(sys.argv[2])
for x in params:
	print(x + ':' + params[x].replace("%2c", ","))

if params["action"] == "search":
  if 'aftermanual' in params:
    details_array = get_details_from_episode_url(params['episodelink'])
    languages = params["languages"].replace("%2c", ",").split(",")
    language_code_string = build_language_code_string(languages)
    show_details = {
      'episode':    details_array[3],
      'season':     details_array[2],
      'showtitle':  details_array[1]
    }
    show_id = get_show_id(show_details['showtitle'])
    showURL = build_show_url(show_id, show_details['season'], language_code_string)
    subs = subs_array(showURL, show_details)
    create_subs_list(subs)
  else:
    languages = params["languages"].replace("%2c", ",").split(",")
    language_code_string = build_language_code_string(languages)
    show_details = get_details_from_player()
    show_id = get_show_id(show_details['showtitle'])
    showURL = build_show_url(show_id, show_details['season'], language_code_string)
    subs = subs_array(showURL, show_details)
    create_subs_list(subs)
elif params["action"] == "download":
  fname = params["filename"].replace(" ", "_") + ".srt"
  subs = download(params["link"], fname)
  for sub in subs:
    listitem = xbmcgui.ListItem(label=sub)
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=sub, listitem=listitem, isFolder=False)
elif params["action"] == "manualsearch":
  episodes = episodes_array(params['searchstring'])
  create_episodes_list(episodes, params['languages'], params['preferredlanguage'])


xbmcplugin.endOfDirectory(int(sys.argv[1]))



resources/lib/koditalk.py
# -*- coding: utf-8 -*-
"""Docstring"""

import json
import logging
import xbmc

def pretty_json(ugly_json):
    """Docstring"""
    return json.dumps(json.loads(ugly_json), indent=2, sort_keys=True)

def get_details_from_player():
    """Docstring"""
    logging.debug("addictedutils.get_details_from_player() begin")
    # Create request
    request = {
        "jsonrpc": "2.0",
        "method": "Player.GetItem",
        "params": {
            "properties": [
                "showtitle",
                "season",
                "episode",
                "duration",
                "file"],
            "playerid": 1
        },
        "id": "VideoGetItem"
    }
    # Make request
    response = xbmc.executeJSONRPC(json.dumps(request))
    logging.debug("response: " + pretty_json(response))
    parsed_response = json.loads(response)['result']['item']
    logging.debug("parsed_response: " + json.dumps(parsed_response, indent=2, sort_keys=True))
    return parsed_response


resources/lib/kodiutils.py
# -*- coding: utf-8 -*-

import xbmc
import xbmcaddon
import re
import sys
import logging
import json as json


# read settings
ADDON = xbmcaddon.Addon()

logger = logging.getLogger(__name__)


def notification(header, message, time=5000, icon=ADDON.getAddonInfo('icon'), sound=True):
    xbmcgui.Dialog().notification(header, message, icon, time, sound)


def show_settings():
    ADDON.openSettings()


def get_setting(setting):
    return ADDON.getSetting(setting).strip().decode('utf-8')


def set_setting(setting, value):
    ADDON.setSetting(setting, str(value))


def get_setting_as_bool(setting):
    return get_setting(setting).lower() == "true"


def get_setting_as_float(setting):
    try:
        return float(get_setting(setting))
    except ValueError:
        return 0


def get_setting_as_int(setting):
    try:
        return int(get_setting_as_float(setting))
    except ValueError:
        return 0


def get_string(string_id):
    return ADDON.getLocalizedString(string_id).encode('utf-8', 'ignore')


def kodi_json_request(params):
    data = json.dumps(params)
    request = xbmc.executeJSONRPC(data)

    try:
        response = json.loads(request)
    except UnicodeDecodeError:
        response = json.loads(request.decode('utf-8', 'ignore'))

    try:
        if 'result' in response:
            return response['result']
        return None
    except KeyError:
        logger.warn("[%s] %s" %
                    (params['method'], response['error']['message']))
        return None


resources/lib/addictedutils.py
# -*- coding: utf-8 -*-
"""Docstring"""

import logging
import requests

from bs4 import BeautifulSoup

BASE_URL = "http://www.addic7ed.com"

def get_language_code(language):
    """Docstring"""
    switcher = {
        "English":    "1",
        "Spanish":    "4",
        "Italian":    "7",
        "French":     "8",
        "Portoguese": "9",
        "Turkish":    "16",
        "Dutch":      "17",
        "Russian":    "19",
        "Romanian":   "26",
        "Greek":      "27",
        "Bulgarian":  "35",
        "Arabic":     "38"
    }
    return switcher.get(language, '')

def build_language_code_string(languages):
    """Docstring"""
    code_string = "|"
    for language in languages:
        code = get_language_code(language)
        if code != "":
            code_string += code + "|"

    if code_string == "|":
        code_string = "|1|"

    return code_string

def build_show_url(show_id, show_season, language_code_string):
    """Docstring"""
    logging.debug("addictedutils.build_show_url() BEGIN")
    show_url = BASE_URL + "/ajax_loadShow.php?"
    show_url += "show=" + str(show_id)
    show_url += "&season=" + str(show_season)
    show_url += "&langs=" + str(language_code_string)
    logging.debug("show_url: " + show_url)
    return show_url

def get_show_id(show_name):
    """Docstring"""
    logging.debug("addictedutils.get_show_id() BEGIN")
    # Get source of addic7ed home page
    source = requests.get(BASE_URL).content.decode('utf-8')
    # logging.debug(source)
    # Trim source to avoid bad parsing
    string_start = "<span id=\"qssShow\">"
    string_end = "<span id=\"qsSeason\">"
    tag_index_start = source.find(string_start)
    tag_index_end = source.find(string_end)
    source = source[tag_index_start:tag_index_end]
    # logging.debug(source)
    # Make the soup
    soup = BeautifulSoup(source, "html.parser")
    show_tag = soup.find('option', text=show_name)
    logging.debug(show_tag)
    logging.debug("show id: " + show_tag['value'])
    return show_tag['value']

def subs_array(show_url, show_details):
    """Docstring"""
    logging.debug("addictedutils.subs_array() BEGIN")
    # Get source
    source = requests.get(show_url).content.decode('utf-8')
    # logging.debug(source)

    # Make the soup
    soup = BeautifulSoup(source, "html.parser")
    # logging.debug("Soup served")

    # Iterate through rows of first table
    # and generate array with subs details
    subs = []
    # logging.debug(show_details['episode'])
    for row in soup.find('table').tbody.findAll('tr'):
        if (not row.has_attr("height")) and (row.contents[1].text == str(int(show_details['episode']))):
            logging.debug("got in")
            if row.contents[7].text.encode('UTF-8') != u'':
                himp = True
            else:
                himp = False
            if row.contents[8].text.encode('UTF-8') != u'':
                sync = True
            else:
                sync = False
            sub = {
                "season": row.contents[0].text,
                "episode": row.contents[1].text,
                "episodeTitle": row.contents[2].contents[0].text,
                "showTitle": show_details['tvshow'],
                "lang":	row.contents[3].text,
                "version": row.contents[4].text,
                "himp": himp,
                "sync": sync,
                "link": row.find_all('a')[1].get('href'),
            }
            subs.append(sub.copy())
    # logging.debug(subs)
    return subs

def download_subtitle(link):
    """Docstring"""
    url = BASE_URL + link
    rheaders = {
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Accept-Encoding': 'gzip, deflate, sdch',
        'Connection': 'keep-alive',
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2715.0 Safari/537.36',
        'Host': 'www.addic7ed.com',
        'Referer': 'http://www.addic7ed.com'}

    response = requests.get(url, headers=rheaders)
    return response.text


resources/lib/subtitle.py
# -*- coding: utf-8 -*-

import urllib2
import requests
import sys
import urlparse
import urllib
import os
import unicodedata
import re
import logging
import codecs
import xbmcgui
import xbmcplugin
import xbmc
import xbmcaddon
import xbmcvfs

import addictedutils

ADDON = xbmcaddon.Addon()
SCRIPT_ID = ADDON.getAddonInfo('id')
PROFILE = xbmc.translatePath(ADDON.getAddonInfo('profile'))
TEMP = os.path.join(PROFILE, 'temp', '')
HANDLE = int(sys.argv[1])

# Create the folder where we will store the downloaded subtitles
if not xbmcvfs.exists(TEMP):
    xbmcvfs.mkdirs(TEMP)

def get_params():
    """ Function to retrieve passed parameters, returns a dict.

        Returns:
            The passed parameters as a dictionary.
    """
    if len(sys.argv) > 2:
        return dict(urlparse.parse_qsl(sys.argv[2].lstrip('?')))
    return {}

def normalize_string(string):
    """Docstring"""
    return unicodedata.normalize(
        'NFKD', unicode(unicode(string, 'utf-8'))
    ).encode('ascii', 'ignore')

def get_info():
    """ Gather all the information available from the playing item,
        helps to determine what subtitle to list
    """
    item = {}
    item['temp'] = False
    item['rar'] = False
    # Year
    item['year'] = xbmc.getInfoLabel("VideoPlayer.Year")
    # Season
    item['season'] = str(xbmc.getInfoLabel("VideoPlayer.Season"))
    # Episode
    item['episode'] = str(xbmc.getInfoLabel("VideoPlayer.Episode"))
    # Show title
    item['tvshow'] = normalize_string(
        xbmc.getInfoLabel("VideoPlayer.TVshowtitle"))
    # try to get original title
    item['title'] = normalize_string(
        xbmc.getInfoLabel("VideoPlayer.OriginalTitle"))
    # Full path of a playing file
    item['file_original_path'] = urllib.unquote(
        xbmc.Player().getPlayingFile().decode('utf-8'))

    if item['title'] == "":
        # no original title, get just Title
        item['title'] = normalize_string(xbmc.getInfoLabel("VideoPlayer.Title"))
        logging.debug(item['title'])
        logging.debug(item['season'])
        logging.debug(item['episode'])
        # If you didn't get any season, episode or tvshow before try and get it from title.
        regex = re.search(r"\bS(\d.)+E(\d+)\b", item['title'])
        (season, episode) = regex.group(1, 2) if (regex is not None) else ("", "")
        item['season'] = season if item['season'] == '' else item['season']
        item['episode'] = episode if item['episode'] == '' else item['episode']
        if item['tvshow'] == '':
            item['tvshow'] = re.sub(r"\b(S\d.+E\d+)\b", '', item['title']).strip()


    # Check if season is "Special"
    if item['episode'].lower().find("s") > -1:
        item['season'] = "0"
        item['episode'] = item['episode'][-1:]

    # Check if the path is not a local one
    if item['file_original_path'].find("http") > -1:
        item['temp'] = True

    # Check if the path is to a rar file
    elif item['file_original_path'].find("rar://") > -1:
        item['rar'] = True
        item['file_original_path'] = os.path.dirname(
            item['file_original_path'][6:])

    # Check if the path is part of a stack of files
    elif item['file_original_path'].find("stack://") > -1:
        stack_path = item['file_original_path'].split(" , ")
        item['file_original_path'] = stack_path[0][8:]

    # Filename
    item['filename'] = os.path.splitext(
        os.path.basename(item['file_original_path']))[0]
    return item

def get_languages(params, lang_format = 2):
    """
    Get the requested languages the user want to search (3 letter format)

    Args:
        params: The passed parameters dict.
        lang_format: 0 to get full language name
            1 to get two letter code (ISO 639-1)
            2 to get three letter code (ISO 639-2/T or ISO 639-2/B) (default)

    Returns:
        An array with the requested languages, e.g. ['scc','eng']
    """
    langs = []  # ['scc','eng']
    for lang in urllib.unquote(params['languages']).decode('utf-8').split(","):
        if lang_format == 0:
            # Full language name
            langs.append(lang)
        elif lang_format == 1:
            # 2 letter format
            langs.append(xbmc.convertLanguage(lang, xbmc.ISO_639_1))
        else:
            # 3 letter format
            langs.append(xbmc.convertLanguage(lang, xbmc.ISO_639_2))
    return langs

def append_subtitle(subname, lang_name, language, params, sync=False, h_impaired=False):
    """Add the subtitle to the list of subtitles to show"""
    listitem = xbmcgui.ListItem(
        # Languange name to display under the lang logo (for example english)
        label=lang_name,
        # Subtitle name (for example 'Lost 1x01 720p')
        label2=subname,
        # Languange 2 letter name (for example en)
        thumbnailImage=xbmc.convertLanguage(language, xbmc.ISO_639_1))

    # Subtitles synced with the video
    listitem.setProperty("sync", 'true' if sync else 'false')
    # Hearing impaired subs
    listitem.setProperty("hearing_imp", 'true' if h_impaired else 'false')

    # Create the url to the plugin that will handle the subtitle download
    url = "plugin://{url}/?{params}".format(
        url=SCRIPT_ID, params=urllib.urlencode(params))
    # Add the subtitle to the list
    xbmcplugin.addDirectoryItem(
        handle=HANDLE, url=url, listitem=listitem, isFolder=False)

def search(info, languages):
    """Add subtitles to the list using information gathered with get_info and get_languages"""
    logging.debug("---TUCAN---")
    logging.debug(info)
    url = addictedutils.build_show_url(
        addictedutils.get_show_id(info['tvshow']),
        info['season'],
        addictedutils.build_language_code_string(languages))
    logging.debug(url)
    subs = addictedutils.subs_array(url, info)
    #logging.debug(subs)
    for sub in subs:
        sub_name = sub['showTitle'] + " " + sub['season'] + "x" + sub['episode'] + " " + sub['version']
        sub_params = {
            "action": "download-addicted",
            "link": sub['link'],
            "name": sub_name
        }
        append_subtitle(
            sub_name,
            sub['lang'],
            xbmc.convertLanguage(sub['lang'], xbmc.ISO_639_2),
            sub_params,
            sub['sync'],
            sub['himp'])

    #addictedutils.get_details_from_player()

def manual_search(search_str, languages):
    """Add subtitles to the list using user manually inserted string"""
    append_subtitle(
        "Lost 1x01", "English", "eng", {"action": "download", "id": 15}, sync=True)
    append_subtitle("Lost 1x01", "Italian", "ita", {"action": "download", "id": 16})
    append_subtitle("Lost 1x01 720p", "English", "eng", {"action": "download", "id": 17})

def download(params):
    """Download the subtitle chosen by the user"""
    id = params['id']
    # download the file requested
    url = "http://path.to/subtitle/{id}.srt".format(id=id)
    file = os.path.join(TEMP, "{id}.srt".format(id=id))

    response = urllib2.urlopen(url)
    with codecs.open(file, "w", "utf-8") as local_file:
        local_file.write(response.read())

    # give the file to kodi
    xbmcplugin.addDirectoryItem(
        handle=HANDLE, url=file, listitem=xbmcgui.ListItem(label=file), isFolder=False)

def download_addicted(params):
    """Download the subtitle from addic7ed.com"""
    file = os.path.join(TEMP, params['name'] + ".srt")
    dfile = addictedutils.download_subtitle(params['link'])

    with open(file, 'wb',) as local_file:
        local_file.write(dfile.encode('utf-8'))

    # Give the file to kodi
    xbmcplugin.addDirectoryItem(
        handle=HANDLE, url=file, listitem=xbmcgui.ListItem(label=file), isFolder=False)

def run():
    """Docstring"""
    # Gather the request info
    params = get_params()

    if 'action' in params:
        if params['action'] == "search":
            # If the action is 'search' use item information kodi provides to search for subtitles
            search(get_info(), get_languages(params, 0))
        elif params['action'] == "manualsearch":
            # If the action is 'manualsearch' use user manually inserted string to search
            # for subtitles
            manual_search(params['searchstring'], get_languages(params, 0))
        elif params['action'] == "download":
            # If the action is 'download' use the info provided to download the subtitle and give
            # the file path to kodi
            download(params)
        elif params['action'] == "download-addicted":
            # If the action is 'download' use the info provided to download the subtitle and give
            # the file path to kodi
            download_addicted(params)

    xbmcplugin.endOfDirectory(HANDLE)
